stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -0,0 +1,720 @@
+
+use clarity::vm::database::BurnStateDB;
+
+use crate::burnchains::db::BurnchainBlockData;
+use crate::burnchains::db::BurnchainDB;
+use crate::burnchains::db::BurnchainHeaderReader;
+use crate::burnchains::Burnchain;
+use crate::burnchains::BurnchainBlockHeader;
+use crate::chainstate::burn::db::sortdb::SortitionDB;
+use crate::chainstate::burn::BlockSnapshot;
+use crate::chainstate::coordinator::comm::{
+    CoordinatorChannels, CoordinatorCommunication, CoordinatorEvents, CoordinatorNotices,
+    CoordinatorReceivers,
+};
+
+use crate::chainstate::coordinator::{
+    calculate_paid_rewards, dispatcher_announce_burn_ops, BlockEventDispatcher, ChainsCoordinator,
+    Error, OnChainRewardSetProvider, PaidRewards, PoxAnchorBlockStatus, RewardCycleInfo,
+    RewardSetProvider,
+};
+use crate::chainstate::nakamoto::NakamotoChainState;
+use crate::chainstate::stacks::boot::RewardSet;
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::miner::signal_mining_blocked;
+use crate::chainstate::stacks::miner::signal_mining_ready;
+use crate::chainstate::stacks::miner::MinerStatus;
+
+use crate::cost_estimates::CostEstimator;
+use crate::cost_estimates::FeeEstimator;
+
+use crate::monitoring::increment_stx_blocks_processed_counter;
+
+use crate::util_lib::db::Error as DBError;
+
+use stacks_common::types::chainstate::BlockHeaderHash;
+use stacks_common::types::chainstate::BurnchainHeaderHash;
+use stacks_common::types::chainstate::SortitionId;
+use stacks_common::types::chainstate::StacksBlockId;
+use stacks_common::types::StacksEpoch;
+use stacks_common::types::StacksEpochId;
+
+#[cfg(test)]
+pub mod tests;
+
+impl OnChainRewardSetProvider {
+    pub fn get_reward_set_nakamoto(
+        &self,
+        // NOTE: this value is the first burnchain block in the prepare phase which has a Stacks
+        // block (unlike in Stacks 2.x, where this is the first block of the reward phase)
+        current_burn_height: u64,
+        chainstate: &mut StacksChainState,
+        burnchain: &Burnchain,
+        sortdb: &SortitionDB,
+        block_id: &StacksBlockId,
+    ) -> Result<RewardSet, Error> {
+        let registered_addrs =
+            chainstate.get_reward_addresses(burnchain, sortdb, current_burn_height, block_id)?;
+
+        let liquid_ustx = chainstate.get_liquid_ustx(block_id);
+
+        let (threshold, participation) = StacksChainState::get_reward_threshold_and_participation(
+            &burnchain.pox_constants,
+            &registered_addrs[..],
+            liquid_ustx,
+        );
+
+        let cur_epoch =
+            SortitionDB::get_stacks_epoch(sortdb.conn(), current_burn_height)?.expect(&format!(
+                "FATAL: no epoch defined for burn height {}",
+                current_burn_height
+            ));
+
+        if participation == 0 {
+            // no one is stacking.  This is a fatal error.
+            error!("No PoX participation. Aborting.");
+            panic!();
+        }
```

Comment:
Should this actually panic, or should the node just stall?

Reply:
I don't think there's a functional difference?  The absence of PoX necessarily means the node can't do anything more.  I can make it just return an error instead, and have the coordinator runloop handle it by stalling if you prefer.

Reply:
I think I would stall. I would expect the node to crash when there is some unrecoverable error. If this can be fixed by waiting and then someone stacking...then I would stall. If it is an irrecoverable error...panic.